live-build/misc/upgrade-scripts/upgrade-container

```diff
@@ -273,12 +273,79 @@ function run() {
 	systemd-run --machine="$CONTAINER" --quiet --wait --pipe -- "$@"
 }
 
+function get_bootloader_devices() {
+	#
+	# When installing/updating the bootloader during upgrade, we
+	# need to determine which devices are being used as bootloader
+	# devices. We determine this by listing the devices used by the
+	# rpool. Additionally, we have to filter out devices that could
+	# be attached to the rpool, but would never be used for the
+	# bootloader. Finally, we need to strip off any parition
+	# information, since we want to install the bootloader directly
+	# to the device, rather than to a partition of the device.
+	#
+	zpool list -vH rpool |
+		awk '! /rpool|mirror|replacing|spare/ {print $1}' |
+		sed 's/[0-9]*$//'
+}
+
+function convert_to_bootfs_cleanup() {
+	for dir in /dev /proc /run /sys; do
+		umount "/var/lib/machines/${CONTAINER}${dir}" ||
+			warn "failed to umount '$dir' from '$CONTAINER'"
+	done
+}
+
+function convert_to_bootfs() {
```

Comment:
Can you elaborate a bit on what is the effect of this?
It seems to be updating and installing new grub. Also, perhaps `update-grub` would be a better name?

Reply:
> Can you elaborate a bit on what is the effect of this?

It's used to convert a upgrade container, to be used as the boot filesystem. It does this by doing things like: 1. update grub to boot the specified container's filesystem on the next boot, 2. update the container's dataset's mountpoint, etc.

> It seems to be updating and installing new grub. Also, perhaps update-grub would be a better name?

It does more than just updating grub though, it also has to change the mountpoint for the container's filesystem.

I chose this name because it takes an "upgrade container" as input, and converts this to be used as the new boot filesystem. I thought simplying naming it "update-grub" wouldn't be a little misleading.

Also, when the actual script is called, it's called like this:
```
$ .../update-container convert-to-bootfs $CONTAINER
```
i.e. it reads like, "convert the specified container to be the new boot filesystem"

which, to me at least, is better than:
```
$ .../update-container update-grub $CONTAINER
```

With that said, I'm fine naming things to whatever makes the most sense to you and everybody else. If you'd rather the option be named "update-grub" then I can make that change.

Reply:
Alright, I think your explanation makes sense.
It would be helpful to add a comment in the code though.

Reply:
:+1: Sure, I can add more comments.